### ISSUE TITLE
Unable to log in via app

### ISSUE THREAD

**Issue by jpalermo** [home-assistant/home-assistant-iOS#238]
I'm consistently getting `Error: invalid client id or redirect uri` from the iOS app when I try to `Validate and Save Connection Settings`.

Hass.io supervisor: 148
Host System: HassOS 2.10
Home Assistant: 0.89.1

This server is running on the local networking being accessed via wifi. From the same iOS device I can successfully log in via Chrome or Safari.

I've tried numerous variations:
http vs https (including generating a self signed cert for `hassio.local` and putting that into a trusted profile on the device)
hassio.local domain name vs ip
uninstalling/reinstalling the app several times

If I opened the failed login page in Safari, this is what the URL looks like:
`10.0.39.80/frontend_es5/authorize.html?response_type=code&client_id=https://home-assistant.io/iOS&redirect_uri=homeassistant://auth-callback`

**Comment by robbiet480** [comment]
What is `frontend_ds5`

**Comment by jpalermo** [comment]
Sorry, typo, es5. I've updated it.

**Comment by jak** [comment]
I have the same issue, however I don’t use a self signed cert over https.

**Comment by robbiet480** [comment]
Recently had another user come to me with this issue and determined it was due to SSL verification failing because they never installed the required certificates, but that was a very specific issue to their setup and platform. I'd suggest upgrading to the latest version of Home Assistant in which I added some logging around failures to validate the client ID. Python/Home Assistant must be able to successfully run a HTTP GET for the `client_id` URL e.g. https://home-assistant.io/iOS. So make sure that you can run `curl https://home-assistant.io/iOS` successfully and don't have some weird networking setup.

**Comment by jpalermo** [comment]
I upgraded to:
Hass.io supervisor: 152
Host System: HassOS 2.11
Home Assistant: 0.90.2

Behavior didn't seem to change. If I curl the `/iOS` path I do end up with a:
`404: Not Found`

Tried trashing the iOS integration, rebooting, adding it back, rebooting. Same behavior, still 404s.

**Comment by WCapernakie** [comment]
I had this same issue. It was related to DNSEC in my PiHole. Disabling DNSEC fixed the issue.

**Comment by vivekmreddy** [comment]
I have the same issue with the redirect error. Only way I can get it working with the iOS app is to use the legacy authentication. Curl works fine for me and the only add-ons I have installed on HassIO are Configurator, Duck DNS, Mosquito Broker, SSH, and Samba.

**Comment by jpalermo** [comment]
Any thoughts on debugging why the `/iOS` endpoint only returns 404?

**Comment by robbiet480** [comment]
@jpalermo The exact endpoint is `/api/ios` but that endpoint won't return anything for a GET. It will return a 404 if the iOS component isn't loaded. The 2.0 beta doesn't require the iOS component anymore though, it needs the `mobile_app` component, which doesn't have a HTTP route you can use to test its loaded.

**Comment by jpalermo** [comment]
@robbiet480 So I think I have the iOS component loaded, there doesn't seem to be much to do configuration wise, just add it right?:
![image](https://user-images.githubusercontent.com/16798/56336552-c95de880-6155-11e9-9062-8c6126e0b2bb.png)

But `/api/ios` still returns a 404:
![image](https://user-images.githubusercontent.com/16798/56336528-b4815500-6155-11e9-8a37-7883455d4e4e.png)

**Comment by robbiet480** [comment]
As I had said, doing a GET to that URL will return a 404 intentionally as it only supports a POST. 

If you are using the 2.0 beta you only need to have mobile_app in your configuration. 

You are still experiencing the initial issue right, something about invalid client ID? If so, it's a issue local to your **Home Assistant setup** not your iOS app. Your Home Assistant installation can't access https://home-assistant.io/iOS for some reason.

**Comment by robbiet480** [comment]
Sorry I misunderstood what you were cURLing. Your HA instance needs to be able to access the URL exactly as written in my last comment. It's not a URL local to your install.

**Comment by robbiet480** [comment]
IPv6 issues are the normal cause of this. Ensure you can do both `curl -v4 https://home-assistant.io/iOS` and `curl -v6 https://home-assistant.io/iOS` from your Home Assistant instance. If the latter command fails you know it's a IPv6 issue.

**Comment by jpalermo** [comment]
Ahh, I thought you were saying the URL would return an empty body if it was configured, and a 404 if it was not configured (and I totally missed the fact that you said `home-assistant.io` and not `hassio.local`)

Does look like it's a IPv6 issue. The curl works from my workstation in both IPv4 and IPv6, but only IPv4 from the hass.io host. Not totally sure why that would be, but I'll dig around the networking config a bit.

**Comment by robbiet480** [comment]
You can just go ahead and safely disable IPv6 entirely on your HA instance. HA will fallback automatically to v4.

**Comment by jpalermo** [comment]
Took a bit of trial and error, not super familiar with NetworkManager but it seems to be working now.

I had tried setting `ipv6.method` to `ignore`, and that might have worked but I didn't check the iOS app when I did that. I was testing the `curl -v6` command, and that was failing.

Eventually I just edited the network config file and removed the whole `ipv6` section. The `curl -v6` still fails, but the iOS app does once again work. Seems a bit strange that the two aren't correlated, but that might just be something weird with the networking on the ssh container I was trying to curl from (curl didn't seem to be available directly on the hass.io host)

Thanks for the help @robbiet480!
